pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/BookkeeperSchemaStorage.java

```diff
@@ -699,15 +705,19 @@ static class LocatorEntry {
         }
     }
 
-    public static Exception bkException(String operation, int rc, long ledgerId, long entryId) {
+    public static Exception bkException(String operation, int rc, long ledgerId, long entryId,
+            boolean forceRecovery) {
         String message = org.apache.bookkeeper.client.api.BKException.getMessage(rc)
                 + " -  ledger=" + ledgerId + " - operation=" + operation;
 
         if (entryId != -1) {
             message += " - entry=" + entryId;
         }
         boolean recoverable = rc != BKException.Code.NoSuchLedgerExistsException
-                && rc != BKException.Code.NoSuchEntryException;
+                && rc != BKException.Code.NoSuchEntryException
+                // if force-recovery is enabled then made it non-recoverable exception
+                // and force schema to skip this exception and recover immediately
+                && !forceRecovery;
```

Comment:
Only for this part of the schema. I had a pr(https://github.com/apache/pulsar/pull/20415) attempt to fix this "unrecoverable errors". The implementation method in PR has not been discussed, but the function should be regarded as completed. So if we have a way to make up for the missing schema, maybe we don't need to skip the "unrecoverable errors" in this part of the schema.

Reply:
as I mentioned it’s very tricky and not possible to conclude non-recoverable errors. For example, the broker can not differentiate between all the ensemble bookies of the ledgers that are temporarily unavailable or are permanently removed from the cluster without graceful recovery.
So, if you always ignore recoverable schema ledgers then it means we lose schema metadata even though ledgers were recoverable. so, it's better to control it with flag and ignore them only if it necessary.